core/states/approaching_gate.py

```diff
@@ -0,0 +1,195 @@
+import asyncio
+from core.vision.ar_tag_detector import is_gate
+import core
+import interfaces
+import algorithms
+from core.states import RoverState
+import time
+import math
+
+
+class ApproachingGate(RoverState):
+    """
+    Within approaching marker, the rover explicitly follows the spotted marker until it reaches an acceptable distance from the rover, or loses sight of it.
+    """
+
+    def start(self):
+        # Schedule AR Tag detection
+        self.num_detection_attempts = 0
+        self.gate_detection_attempts = 0
+
+    def exit(self):
+        # Cancel all state specific coroutines
+        pass
+
+    def on_event(self, event) -> RoverState:
+        """
+        Defines all transitions between states based on events
+        """
+        state: RoverState = None
+
+        if event == core.AutonomyEvents.REACHED_MARKER:
+            state = core.states.Idle()
+
+        elif event == core.AutonomyEvents.START:
+            state = self
+
+        elif event == core.AutonomyEvents.MARKER_UNSEEN:
+            state = core.states.SearchPattern()
+
+        elif event == core.AutonomyEvents.ABORT:
+            state = core.states.Idle()
+
+        else:
+            self.logger.error(f"Unexpected event {event} for state {self}")
+            state = core.states.Idle()
+
+        # Call exit() if we are not staying the same state
+        if state != self:
+            self.exit()
+
+        # Return the state appropriate for the event
+        return state
+
+    async def run(self) -> RoverState:
+
+        # Call AR Tag tracking code to find position and size of AR Tag
+        if core.vision.ar_tag_detector.is_gate():
+            # Grab the AR tags
+            tags = core.vision.ar_tag_detector.get_tags()
```

Comment:
Maybe improve this variable name or comment more thoroughly what get_tags actually returns